pypiserver/config.py

```diff
@@ -0,0 +1,599 @@
+            )
+            try:
+                # Prevent showing the error for the adjusted args, since
+                # we want to show the captured error from the user's
+                # actually provided args.
+                with capture_stderr():
+                    parsed = parser.parse_args(cls._adjust_old_args(args))
+            except SystemExit:
+                # Here we could not parse our adjusted arguments either,
+                # So we'll print the original error message and rethrow
+                # the exit.
+                print(err_msg, file=sys.stderr)
+                raise
+            else:
+                # Here on the other hand we _were_ able to parse the adjusted
+                # arguments, which means the user is using a deprecated form.
+                # Print out a warning, along with our original error or
+                # usage string.
+                print(
+                    "Warning: you are using deprecated cmdline arguments. "
+                    "Please see the usage message below, and update your "
+                    "cmdline arguments.\n",
+                    file=sys.stderr,
+                )
+                print(err_msg, file=sys.stderr)
+
+        if parsed.cmd == "run":
+            return RunConfig(parsed)
+        if parsed.cmd == "update":
+            return UpdateConfig(parsed)
+        raise SystemExit(parser.format_usage())
+
+    @staticmethod
+    def _adjust_old_args(args: t.Sequence[str]) -> t.List[str]:
+        """Adjust args for backwards compatibility.
+
+        Should only be called once args have been verified to be unparseable.
+        """
+        # Backwards compatibility hack: for most of pypiserver's life, "run"
+        # and "update" were not separate subcommands. The `-U` flag being
+        # present on the cmdline, regardless of other arguments, would lead
+        # to update behavior. In order to allow subcommands without
+        # breaking backwards compatibility, we need to insert "run" or
+        # "update" as a positional arg before any other arguments.
+
+        # We will be adding our subcommand as the first argument.
+        insertion_idx = 0
+
+        # Don't actually update the passed list, in case it was the global
+        # sys.argv.
+        args = list(args)
+
+        # Find the update index. For "reasons", python's index search throws
+        # if the value isn't found, so manually wrap in the usual "return -1
+        # if not found" standard
+        try:
+            update_idx = args.index("-U")
+        except ValueError:
+            update_idx = -1
+
+        if update_idx == -1:
+            # We were a "run" command.
+            args.insert(insertion_idx, "run")
+        else:
+            # Remove the -U from the args and add the "update" command at the
+            # start of the arg list.
+            args.pop(update_idx)
+            args.insert(insertion_idx, "update")
+
+        return args
```

Comment:
What happens when someone calls `pypiserver run -U`?

Reply:
argparse would throw, because it parses the `run` subcommand first, and recognizes that `-U` is not a valid argument for that subcommand.

Adding this in to the bottom of this file:

```py
print(Config.from_args(["run", "-U"]))
```

I get:

```
usage: config.py [-h] {run,update} ...
config.py: error: unrecognized arguments: -U
```